src/main-oe/serialize_enclave_config.c

```diff
@@ -0,0 +1,532 @@
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <json.h>
+#include "enclave/enclave_mem.h"
+#include "enclave/wireguard.h"
+#include "host/sgxlkl_util.h"
+#include "shared/env.h"
+#include "shared/host_state.h"
+#include "shared/sgxlkl_enclave_config.h"
+
+#include "host/serialize_enclave_config.h"
+
+#define FAIL sgxlkl_host_fail
+#define VERB sgxlkl_host_verbose
+
+typedef struct json_obj
+{
+    char* key;
+    json_type_t type;
+    union {
+        char* string;
+        bool boolean;
+    } value;
+    struct json_obj** array;
+    struct json_obj** objects;
+    size_t size;
+} json_obj_t;
+
+static json_obj_t* mk_json_obj(
+    const char* key,
+    json_type_t type,
+    json_obj_t** array,
+    json_obj_t** objects,
+    size_t size)
+{
+    json_obj_t* r = calloc(1, sizeof(json_obj_t));
+    if (!r)
+        FAIL("out of memory\n");
+    if (key)
+        r->key = strdup(key);
+    r->type = type;
+    r->value.string = NULL;
+    r->array = array;
+    r->objects = objects;
+    r->size = size;
+    return r;
+}
+
+static json_obj_t* mk_json_string(const char* key, const char* value)
+{
+    json_obj_t* obj = mk_json_obj(key, JSON_TYPE_STRING, NULL, NULL, 0);
+    if (value)
+    {
+        obj->value.string = strdup(value);
+        if (!obj->value.string)
+            FAIL("out of memory\n");
+    }
+    return obj;
+}
+
+static json_obj_t* mk_json_hex_string(
+    const char* key,
+    const uint8_t* buf,
+    size_t buf_sz)
+{
+    json_obj_t* obj = mk_json_obj(key, JSON_TYPE_STRING, NULL, NULL, 0);
+    if (buf && buf_sz > 0)
+    {
+        size_t len = 2 * buf_sz + 1;
+        obj->value.string = malloc(len);
+        if (!obj->value.string)
+            FAIL("out of memory\n");
+        bytes_to_hex(obj->value.string, len, buf, buf_sz);
+    }
+    return obj;
+}
+
+static json_obj_t* mk_json_boolean(const char* key, const bool value)
+{
+    json_obj_t* obj = mk_json_obj(key, JSON_TYPE_BOOLEAN, NULL, NULL, 0);
+    obj->value.boolean = value;
+    return obj;
+}
+
+static json_obj_t* mk_json_array(const char* key, size_t len)
+{
+    json_obj_t* r = mk_json_obj(key, JSON_TYPE_NULL, NULL, NULL, len);
+    r->array = calloc(len, sizeof(char*));
+    if (!r->array)
+        FAIL("out of memory\n");
+    return r;
+}
+
+static json_obj_t* mk_json_objects(const char* key, size_t len)
+{
+    json_obj_t* r = mk_json_obj(key, JSON_TYPE_NULL, NULL, NULL, len);
+    r->objects = calloc(len, sizeof(json_obj_t*));
+    if (!r->objects)
+        FAIL("out of memory\n");
+    return r;
+}
+
+static json_obj_t* mk_json_int(const char* key, uint64_t value, const char* fmt)
+{
+    size_t len = 8 * 2 + 1;
+    char* tmp = calloc(1, len);
+    if (!tmp)
+        FAIL("out of memory\n");
+    snprintf(tmp, len, fmt, value);
+    return mk_json_string(key, tmp);
+}
+
+static json_obj_t* mk_json_u32(const char* key, uint32_t value)
```

Comment:
The word "canonical" has a precise mathematical definition, which I believe our JSON encoding does not meet, so I'd rather not use that term. I'll rename the functions though, `mk_` is not a convention in this code base.

Reply:
If there's anything non-canonical with what we have then we need to fix it, otherwise attestation becomes pointless :) Or do you mean something else?

Reply:
I don't think we ever finished that discussion and just accepted the fact that JSON will never be truly canonical. For instance, strings would still allow leading zeros (although I think I disallowed them in the schema) and there may be an optional `+` sign. Also, the order of array elements is not necessarily fixed (I think it's fine in our case, but I'm not sure). I don't even want start thinking about Unicode (in the JSON standard). Then there are a bunch of issues with intentionally mis-constructed files. All these problems can probably be address in some way, but the right way to make sure is to prove it (e.g. let EverParse produce a verified parser for our subset of JSON) but nobody has time to do that at the moment.

Reply:
These are important points, please don't keep them to yourself. It's much more useful if you open an issue for it so that we are at least aware of them.

Reply:
Thanks for reminding me about that! Added them to the existing issue on that topic.